src/stores/A_store.js

```diff
@@ -60,15 +63,89 @@ class A_store extends ReduceStore {
-                return {pathGrid: new Grid(SIZE, SIZE), player: {x: -1, y: -1}, end: {x: -2, y: -2}, stage: 'STARTP', path: ''};
+                return {
+                    pathGrid: new Grid(SIZE, SIZE), 
+                        player: {x: -1, y: -1},
+                        end: {x: -2, y: -2}, 
+                        stage: 'STARTP', 
+                        path: null, 
+                        visibleCells: null, 
+                        seenCells: null
+                    };
             }
 
             case ActionTypes.STEP: {
-                const path = [...state.path];
-                const nextLocation = path.shift();
+                const [nextLocation, ...path] = state.path;
 
                 const stage = path.length > 0 ? 'STEP' : 'RESET';
-                return {...state, stage, path, player: {x: nextLocation.x, y: nextLocation.y}};
+                return {
+                    ...state, 
+                        stage, 
+                        path, 
+                        player: {x: nextLocation.x, y: nextLocation.y}
+                    };
+            }
+
+            case ActionTypes.GENERATE_FOG: {
+                const {player, end} = state;
+                const {x, y} = player;
+                const revealedGrid = mapGenerator(player, end, new Grid(SIZE, SIZE));
+                const pathGrid = new Grid(SIZE, SIZE);
+
+                const visibleCells = pathGrid.getVisible(x, y);
+                revealedGrid.copyCells(visibleCells);
+                const seenCells = pathGrid.getVisible(x, y);
+
+                const start = pathGrid.getCell(x, y);
+                const goal = pathGrid.getCell(end.x, end.y);
+                const path = Astar(start, goal, Cell.heuristic);
+                path.shift();
+
+                return {
+                        ...state, 
+                        pathGrid, 
+                        revealedGrid,
+                        path, 
+                        stage: 'STEP_FOG', 
+                        visibleCells, 
+                        seenCells
+                    };
+            }
+
+            case ActionTypes.STEP_FOG: {
+                let [nextLocation, ...path] = state.path;
+                const {revealedGrid, end, seenCells} = state;
+                let {pathGrid} = state;
+                let stage;
+                let visibleCells;
+
+                if (path.length > 0) {
+                    stage = 'STEP_FOG';
+                    visibleCells = pathGrid.getVisible(nextLocation.x, nextLocation.y);
+                    revealedGrid.copyCells(visibleCells);
+
+                    for (const cell of visibleCells) {
+                        seenCells.add(cell);
+                    }
+
+                    const start = pathGrid.getCell(nextLocation.x, nextLocation.y);
+                    const goal = pathGrid.getCell(end.x, end.y);
+                    path = Astar(start, goal, Cell.heuristic);
+                    path.shift();
```

Comment:
so, i'm noticing something here

this is `GENERATE_FOG`:
```js
                const visibleCells = pathGrid.getVisible(x, y);
                revealedGrid.copyCells(visibleCells);

                const seenCells = pathGrid.getVisible(x, y); // same as visibleCells

                const start = pathGrid.getCell(x, y);
                const goal = pathGrid.getCell(end.x, end.y);
                const path = Astar(start, goal, Cell.heuristic);
                path.shift();
```

this is `STEP_FOG`:
```js
                    visibleCells = pathGrid.getVisible(nextLocation.x, nextLocation.y);
                    revealedGrid.copyCells(visibleCells);

                    for (const cell of visibleCells) {
                        seenCells.add(cell);
                    }

                    const start = pathGrid.getCell(nextLocation.x, nextLocation.y);
                    const goal = pathGrid.getCell(end.x, end.y);
                    path = Astar(start, goal, Cell.heuristic);
                    path.shift();
```
feeling like there is some DRY opportunity here

i don't think this is a blocker tho